### ISSUE TITLE
How to save a trained model

### ISSUE THREAD

**Issue by darkdark87** [lucasmaystre/kickscore#4]
Hello,

I've tried saving the trained model using `pickle` and sklearn `joblib`. Both breaks down with `maximum recursion depth`-error. I've tried raising the default level, but the pickle job runs in what seems forever, without terminating.

Do you have a preferred method for saving a trained model?

Thanks for an excellent library!

**Comment by lucasmaystre** [comment]
Hi @darkdark87 , thanks for the kind words!

I've actually never had to save models before. I just tried using pickle now at the end of this [colab notebook](https://colab.research.google.com/github/lucasmaystre/kickscore/blob/master/examples/nba-history.ipynb):

```
import pickle

with open("test.pkl", "wb") as f:
    pickle.dump(model, f)

with open("test.pkl", "rb") as f:
    m2 = pickle.load(f)

m2.plot_scores(
        items=["LAL", "CHI", "BOS"],
        resolution=10/seconds_in_year,
        figsize=(14.0, 3.0),
        timestamps=True)
```

and it seems to work fine. Could you share a minimum working example where it fails?

**Comment by darkdark87** [comment]
Hi @lucasmaystre 

I believe the errors lies with the number of teams. If I increase the number of teams by a factor 3 I receive an error when pickling the model.

I created a minimum "working example" in [this notebook](https://colab.research.google.com/drive/1zgiqeCcaG2Pdvm7WMYaEjb1gKNmkGT4n).

**Comment by lucasmaystre** [comment]
Thanks, the example was helpful. I think I understand where the problem—there are circular references between `Item` and `Observation`, and I understand it trips pickle up after a certain point.

Putting this on my TODO list, but it'll probably have to wait for a few weeks.

In the meantime, if all you're interested in doing with the reloaded model is to make predictions, the following should work reasonably well:

```
import pickle

# Saving.
fitters = {name: val.fitter for name, val in model.item.items()}
with open("test1.pkl", "wb") as f:
    pickle.dump(fitters, f)

# Re-loading.
with open("test1.pkl", "rb") as f:
    fitters = pickle.load(f)

m2 = ks.BinaryModel()
dummy = ks.kernel.Constant(var=1.0)

for name, fitter in fitters.items():
    m2.add_item(name, kernel=dummy)
    m2.item[name].fitter = fitter
```

It only saves the parameters of the model (which are in the items' `fitter` property) though, so you won't be able to retrain it, or inspect the observations it was trained on, etc. However, `model.probabilites()`, `item.predict()` should work fine.

I'll leave the issue open until I fix persistency in a cleaner way.

**Comment by darkdark87** [comment]
Sounds good @lucasmaystre 
My use case requires retraining of the persisted model, so I'll wait.

Thank you.

**Comment by lucasmaystre** [comment]
Hi @darkdark87 I just released a new version with a fix (v0.1.5). Hopefully it works now—let me know if you still have issues.
